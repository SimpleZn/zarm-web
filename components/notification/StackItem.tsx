import React from 'react';
import Transition from '../transition';
import { NotificationStackItemProps } from './PropsType';

export default class StackItem extends React.Component<NotificationStackItemProps, {}> {
  private timeout: number;

  static defaultProps = {
    stayTime: 5000,
  };

  state = { visible: false };

  componentDidMount() {
    this.setState({ visible: true });
    this.startTimer();
  }

  componentDidUpdate(prevProps) {
    const { stayTime } = prevProps;
    const { stayTime: newTime } = this.props;
    if (stayTime !== newTime) {
      this.stopTimer();
      this.startTimer();
    }
  }

  componentWillUnmount() {
    clearInterval(this.timeout);
  }

  close = () => {
    this.stopTimer();
    this.setState({ visible: false }, this.props.onClose);
  };

  startTimer = () => {
    const { stayTime } = this.props;
    if (stayTime) {
      this.timeout = setTimeout(this.close, stayTime);
    }
  };

  stopTimer = () => {
    clearInterval(this.timeout);
  };

  render() {
    const { Component, name, willUnmount, stayTime, onClose, ...rest } = this.props;
    const { visible } = this.state;
    return (
      <Transition
        visible={visible}
        name={name}
        duration={300}
        unmountOnHide
        onHide={willUnmount}
      >
        <Component
          {...rest}
          onMouseEnter={this.stopTimer}
          onMouseLeave={this.startTimer}
          onClose={this.close}
        />
      </Transition>
    );
  }
}
