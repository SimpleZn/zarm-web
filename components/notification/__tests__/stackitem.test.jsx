import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import StackItem from '../StackItem';

function TempComponent(props) {
  const { onClose, test } = props;
  return <div onClick={onClose}>{test}</div>;
}

describe('StackItem', () => {
  it('renders basic StackItem correctly', () => {
    const wrapper = mount(
      <StackItem
        Component={TempComponent}
        name="testname"
        stayTime={4500}
        test="1"
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.text()).toBe('1');
  });

  it('triggers onClose callback correctly', () => {
    const onClose = jest.fn();
    const wrapper = mount(
      <StackItem
        Component={TempComponent}
        name="testname"
        stayTime={4500}
        onClose={onClose}
        test="1"
      />,
    );
    wrapper.find('div').simulate('click');
    expect(onClose).toBeCalled();
  });

  it('start timer correctly', () => {
    jest.useFakeTimers();
    mount(
      <StackItem
        Component={TempComponent}
        name="testname"
        stayTime={1234}
      />,
    );
    expect(setTimeout).toHaveBeenCalledWith(expect.any(Function), 1234);
  });

  it('do not start timer', () => {
    jest.useFakeTimers();
    mount(
      <StackItem
        Component={TempComponent}
        name="testname"
        stayTime={0}
      />,
    );
    // 除了 transition 调了一次，不会有第二次
    expect(setTimeout).toHaveBeenCalledTimes(1);
  });

  it('end timer correctly', () => {
    jest.useFakeTimers();
    const onClose = jest.fn();
    mount(
      <StackItem
        Component={TempComponent}
        name="testname"
        stayTime={1234}
        onClose={onClose}
      />,
    );
    jest.runAllTimers();
    expect(onClose).toBeCalled();
  });

  it('update timer correctly', () => {
    jest.useFakeTimers();
    const onClose = jest.fn();
    const props = {
      Component: TempComponent,
      name: 'testname',
      stayTime: 0,
      onClose,
    };
    const wrapper = mount(
      <StackItem {...props} />,
    );
    expect(onClose).toHaveBeenCalledTimes(0);
    setTimeout(() => {
      wrapper.setProps({
        ...props,
        stayTime: 1500,
      });
    }, 1000);
    jest.runAllTimers();
    expect(onClose).toHaveBeenCalledTimes(1);
  });

  it('unount Component correctly', () => {
    jest.useFakeTimers();
    const wrapper = mount(
      <StackItem
        Component={TempComponent}
        name="testname"
        stayTime={1234}
      />,
    );
    wrapper.unmount();
    expect(clearTimeout).toBeCalled();
  });
});
