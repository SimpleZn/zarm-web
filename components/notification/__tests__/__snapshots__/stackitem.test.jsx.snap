// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StackItem renders basic StackItem correctly 1`] = `
<StackItem
  Component={[Function]}
  name="testname"
  stayTime={4500}
  test="1"
>
  <Transition
    directional={true}
    duration={300}
    mountOnShow={true}
    name="testname"
    onBeforeHide={[Function]}
    onBeforeShow={[Function]}
    onComplete={[Function]}
    onHide={[Function]}
    onShow={[Function]}
    onStart={[Function]}
    transitionOnMount={false}
    unmountOnHide={true}
    visible={true}
  >
    <TempComponent
      className="animating testname testname-in visible transition"
      onClose={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "animationDuration": "300ms",
        }
      }
      test="1"
    >
      <div
        onClick={[Function]}
      >
        1
      </div>
    </TempComponent>
  </Transition>
</StackItem>
`;
